Cython/Compiler/ModuleNode.py

```diff
@@ -1596,45 +1596,64 @@ def generate_new_function(self, scope, code, cclass_entry):
+                code.putln(f"if (likely((int)({code.name_in_slot_module_state(freecount_name)} > 0)) &")
+                code.putln("#if CYTHON_USE_TYPE_SPECS")
+                # with CYTHON_USE_TYPE_SPECS we can only reasonably use freelists for an exact type match,
+                # because everything fails the heap-type check.
+                code.putln(f"(t == {code.name_in_slot_module_state(type.typeptr_cname)}){heap_type_safety_check}")
+                code.putln("#else")
+                code.putln(f"(int)(t->tp_basicsize == sizeof({obj_struct})){type_safety_check}")
+                code.putln("#endif")
+                code.putln(") {")
                 code.putln("o = (PyObject*)%s[--%s];" % (
                     code.name_in_slot_module_state(freelist_name),
                     code.name_in_slot_module_state(freecount_name)))
                 code.putln("memset(o, 0, sizeof(%s));" % obj_struct)
+                code.putln("#if CYTHON_COMPILING_IN_LIMITED_API")
+                # Although PyObject_INIT should be part of the Limited API, it causes
+                # link errors on some combinations of Python versions and OSs.
+                code.putln("(void) PyObject_Init(o, t);")
+                code.putln("#else")
                 code.putln("(void) PyObject_INIT(o, t);")
+                code.putln("#endif")
                 if scope.needs_gc():
                     code.putln("PyObject_GC_Track(o);")
                 code.putln("} else")
                 code.putln("#endif")
                 code.putln("{")
             if not is_final_type:
                 code.putln("if (likely(!__Pyx_PyType_HasFeature(t, Py_TPFLAGS_IS_ABSTRACT))) {")
-            code.putln("o = (*t->tp_alloc)(t, 0);")
+            code.putln("allocfunc alloc_func = __Pyx_PyType_GetSlot(t, tp_alloc, allocfunc);")
+            code.putln("o = alloc_func(t, 0);")
             if not is_final_type:
                 code.putln("} else {")
-                code.putln("o = (PyObject *) PyBaseObject_Type.tp_new(t, %s->%s, 0);" % (
+                code.putln("newfunc tp_new = __Pyx_PyType_TryGetSlot(&PyBaseObject_Type, tp_new, newfunc);")
+                code.putln("#if CYTHON_COMPILING_IN_LIMITED_API && __PYX_LIMITED_VERSION_HEX < 0x030A0000")
+                code.putln("if (!tp_new) {")
+                code.putln('PyObject *new_str = PyUnicode_FromString("__new__");')
+                code.putln("if (likely(new_str)) {")
+                code.putln("o = PyObject_CallMethodObjArgs((PyObject *)&PyBaseObject_Type, new_str, t, NULL);")
+                code.putln("Py_DECREF(new_str);")
+                code.putln("} else")
+                code.putln("o = NULL;")
+                code.putln("} else")
+                code.putln("#endif")
+                code.putln("o = tp_new(t, %s->%s, 0);" % (
                     Naming.modulestateglobal_cname, Naming.empty_tuple))
```

Comment:
I found it hard to follow the if-else chaining here, but this seems a sufficiently long constant piece of code to move it into the utility code files, which would solve that problem.